switch from WCSLIB to WCS (requires Julia 0.4)

kbarbary opened the pull request:
Start using [WCS](http://github.com/JuliaAstro/WCS.jl) instead of WCSLIB, which is deprecated and stuck on an old version of the C wcslib library. This requires renaming the WCS module in this package, because the modules are placed on Julia's load path. I chose the name WCSApprox, somewhat haphazardly! Open to other suggestions.

Of course, Celeste will have to be updated to reflect this change, which I can do once the rename is settled.

rgiordan commented on src/WCSApprox.jl:
```diff
@@ -1,14 +1,15 @@
+"""
+Approximations for WCS transformations.
+"""
+module WCSApprox
```
I think WCSApprox is not a good name.  I'm not sure what is approximate about this module.  How about WCS_SDSS?

kbarbary commented:
The module seems to contain mostly local linearized approximations to WCS transforms. Is that not right?

We could leave the name as `WCS` if we make it a submodule of `SloanDigitalSkySurvey`. I think this might get a bit confusing though.

Unless I'm missing something, nothing in the module is actually specific to SDSS, so perhaps it should just be moved to Celeste?

rgiordan commented:
The point of this library was to abstract out functions that might be useful to everyone processing the SDSS data in julia from functions that are Celeste-specific.  I did it because some friends said they were interested in processing SDSS data in julia and Celeste has too many unnecessary dependencies.  I don't see why it should be part of the Celeste package except for logistical reasons related to maintenance that can be just as easily fixed by making you guys contributors.

As-is, the WCS library contains much more than just the linear approximation (e.g. ```pixel_to_world```, etc.).  It's true that most of what is not a linear approximation is now redundant with your new WCS stuff and could be deleted, but until that's done I don't want to give it a name that does not describe what it is.  Sorry, I aspire to be a bit finicky about naming things.  Either a sub-module or a more descriptive name is fine with me.

kbarbary commented:
I appreciate name finickiness! I'll try the submodule route.

kbarbary commented:
quick question: are the `DualNumber`-related methods in SloanDigitalSkySurvey.WCS still necessary, or should I remove them now?

rgiordan commented:
They should not be necessary, cut 'em!

On Sat, Feb 6, 2016 at 2:53 PM, Kyle Barbary <notifications@github.com>
wrote:

> quick question: are the DualNumber-related methods in
> SloanDigitalSkySurvey.WCS still necessary, or should I remove them now?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rgiordan/SloanDigitalSkySurvey.jl/pull/4#issuecomment-180879197>
> .
>

kbarbary commented:
Two remaining questions here:

1. Submodules aren't enough to avoid a rename of `WCS`. How about `WCSUtils` or maybe `WCSExtensions`?

2. I couldn't quite mentally parse the convention you were using in `SloanDigitalSkySurvey.WCS.world_to_pixel/pixel_to_world` for the axis order of the input coordinates. In `WCS`/`WCSLIB` the convention is that `coords[:, 1]` is the first set of coordinates, so for a 2-d image, `coords` has shape `(2, N)` where `N` is the number of coordinates. It seems like it was opposite in this module?

rgiordan commented:
>
>
>    1.
>
>    Submodules aren't enough to avoid a rename of WCS. How about WCSUtils
>    or maybe WCSExtensions?
>
> Either is fine with me.


>    1.
>
>    I couldn't quite mentally parse the convention you were using in
>    SloanDigitalSkySurvey.WCS.world_to_pixel/pixel_to_world for the axis
>    order of the input coordinates. In WCS/WCSLIB the convention is that coords[:,
>    1] is the first set of coordinates, so for a 2-d image, coords has
>    shape (2, N) where N is the number of coordinates. It seems like it
>    was opposite in this module?
>
> Yes, it is the opposite in this module.  I'm afraid I can't remember why I
did it that way other than a general principle that rows of an array
usually denote distinct "observations".  If you want to change it, I have
no objection, though it will have to be changed in Celeste, too.

kbarbary commented:
> did it that way other than a general principle that rows of an array usually denote distinct "observations".

That's a reasonable principle. However, I think it is indeed worth changing this to be consistent with WCS. Besides consistency for its own sake, the WCS convention has a couple other nice advantages:

- Works naturally with broadcasting in Julia, when a single point is a Vector:
  ```julia
julia> coords = rand(2, 5)  # coordinates of 5 points in two dimensions
2x5 Array{Float64,2}:
 0.986086  0.427589   0.909639  0.603009  0.308425
 0.437676  0.0745297  0.63629   0.633472  0.19575 

julia> coords .- [0.5, 0.5]  # offset of points from coordinates (0.5, 0.5)
2x5 Array{Float64,2}:
  0.486086   -0.072411  0.409639  0.103009  -0.191575
 -0.0623236  -0.42547   0.13629   0.133472  -0.30425 
```
  This allows us to reduce the Jacobian method to a single line:
  ```julia
wcs_jacobian * (worldcoords .- world_offset) + pix_offset
```
which works when `worldcoords` is 1-d or 2-d.

- Having a single point's coordinates contiguous in memory allows zero-copy reinterpretation as specialized coordinate types:
  ```julia
julia> using SkyCoords
julia> skycoords = reinterpret(ICRSCoords, coords, (5,))
5-element Array{SkyCoords.ICRSCoords,1}:
 SkyCoords.ICRSCoords(0.9860864658256394,0.4376764429103155) 
 SkyCoords.ICRSCoords(0.4275890344842139,0.07452970492855537)
 SkyCoords.ICRSCoords(0.9096385242870282,0.6362899337436576) 
 SkyCoords.ICRSCoords(0.6030094716473864,0.6334715060711174) 
 SkyCoords.ICRSCoords(0.3084254930851158,0.19574999831612527)
julia> # ... can then convert to Galactic or other coordinate system.
```

These changes are now implemented in this PR. I've updated Celeste locally to reflect this and will submit a PR there soon.

kbarbary commented:
Want me to submit a new tag to metadata?

kbarbary commented:
OK I'm doing it.

kbarbary commented:
https://github.com/JuliaLang/METADATA.jl/pull/4581
